File: typescript/src/primitives/board.ts
import { Coordinates, UnitType } from '../types.js';
import { isSquarePositiveInteger } from '../utils.js';
import { Cell } from './cell.js';
import { Unit } from './unit.js';

/**
 * Params for constructing a Board
 */
export interface BoardParams {
  cells: Cell[][];
}

/**
 * A complete Sudoku board
 */
export class Board {
  // Box edge size for easy lookup
  readonly boxEdgeSize: number;
  // Unit size for easy lookup
  readonly unitSize: number;
  // Total board size for easy lookupc
  readonly totalBoardSize: number;
  // Cells contained within the board
  readonly cells: Cell[][];
  // Units contained within the board
  readonly units: Unit[];

  constructor(params: BoardParams) {
    const { cells } = params;
    // Validations
    // Check cells 2D array for individual side lengths
    if (!isSquarePositiveInteger(cells.length)) {
      throw new Error(`Height is not a square! Given height: ${cells.length}`);
    }

    let width: number | undefined;
    cells.forEach((row) => {
      if (width !== undefined && width !== row.length) {
        throw new Error(
          'All rows must be the same width! ' +
            `Previous width: ${width} ` +
            `Current width: ${width}`
        );
      }

      if (!isSquarePositiveInteger(row.length)) {
        throw new Error(`Width is not a square! Given width: ${row.length}`);
      }

      if (cells.length !== row.length) {
        throw new Error(
          'Width and height are not the same! Board is not a square! ' +
            `Height: ${cells.length} ` +
            `Width: ${row.length}`
        );
      }

      width = row.length;
    });

    this.boxEdgeSize = Math.sqrt(cells.length);
    this.unitSize = cells.length;
    // Width is guaranteed to be assigned by this point so this non-null assertion is safe
    this.totalBoardSize = cells.length * width!;
    this.cells = cells;
    this.units = this.buildUnits();
  }

  /**
   * Builds units from a 2D array of cells
   * @returns Array of Units (rows, columns, boxes)
   */
  private buildUnits() {
    const rowCoords: Coordinates[][] = Array.from(
      { length: this.unitSize },
      () => []
    );
    const colCoords: Coordinates[][] = Array.from(
      { length: this.unitSize },
      () => []
    );
    const boxCoords: Coordinates[][] = Array.from(
      { length: this.unitSize },
      () => []
    );

    for (let row = 0; row < this.cells.length; row++) {
      for (let col = 0; col < this.cells[row]!.length; col++) {
        const coords: Coordinates = {
          col,
          row,
        };

        const boxRow = Math.floor(row / this.boxEdgeSize);
        const boxCol = Math.floor(col / this.boxEdgeSize);
        const boxIndex = boxRow * this.boxEdgeSize + boxCol;

        // We pre-initialized these arrays, so the non-null accesses are safe
        rowCoords[row]!.push(coords);
        colCoords[col]!.push(coords);
        boxCoords[boxIndex]!.push(coords);
      }
    }

    const units: Unit[] = [];

    for (const row of rowCoords) {
      units.push(new Unit({ unitType: UnitType.Row, cellCoords: row }));
    }

    for (const col of colCoords) {
      units.push(new Unit({ unitType: UnitType.Column, cellCoords: col }));
    }

    for (const box of boxCoords) {
      units.push(new Unit({ unitType: UnitType.Box, cellCoords: box }));
    }

    return units;
  }

  /**
   * Clones the board for use in backtracking
   * @returns A cloned board
   */
  public clone() {
    const clonedCells = this.cells.map((row) =>
      row.map((cell) => cell.clone())
    );
    return new Board({ cells: clonedCells });
  }

  /**'
   * Validates that the current board state is valid
   * @throws A list of board violations
   */
  public validate() {
    const violations = new Map<Unit, Cell[]>();

    this.units.forEach((unit) => {
      const cells = this.getCellsForUnit(unit);

      const duplicatesMap = new Map<number, Cell[]>();
      cells.forEach((cell) => {
        const value = cell.getValue();
        if (value === undefined) {
          return;
        }
        const coordinatesWithValue: Cell[] = duplicatesMap.get(value) ?? [];
        coordinatesWithValue.push(cell);
        duplicatesMap.set(value, coordinatesWithValue);
      });

      duplicatesMap.values().forEach((cells) => {
        if (cells.length > 1) {
          violations.set(unit, cells);
        }
      });
    });

    if (violations.size > 0) {
      const stringifiedViolations = Array.from(
        violations.entries().map((entry) => {
          const stringifiedUnit = entry[0].toString();

          const stringifiedCells = entry[1]
            .map((cell) => cell.toString())
            .join(', ');

          return `{ Violated Unit: ${stringifiedUnit}, Violated Cells: ${stringifiedCells} }`;
        })
      );

      throw new Error(
        `Board is invalid! Violations: [ ${stringifiedViolations.join(', ')} ]`
      );
    }
  }

  /**
   * Determines if every cell in the board has a value. Does not check validity
   * @returns if the board is already filled
   */
  public isFilled() {
    const cellsWithValues = this.cells
      .flat()
      .filter((cell) => cell.getValue() !== undefined);
    return cellsWithValues.length === this.totalBoardSize;
  }

  /**
   * Determines if the board is already solved
   * @returns If the board is solved or not
   */
  public isSolved() {
    try {
      this.validate();
    } catch (error) {
      console.log(error as Error);
      return false;
    }

    return this.isFilled();
  }

  public getCellsForUnit(unit: Unit) {
    return unit.cellCoords.map((coord) => this.cells[coord.row]![coord.col]!);
  }

  /**
   * Get the empty cell with the fewest candidates (best cell for use in backtracking)
   * @returns the empty cell with the fewest candidates
   * @throws if the board is already filled and there are no empty cells
   */
  public selectEmptyCell() {
    const emptyCells = this.cells
      .flat()
      .filter((cell) => cell.getValue() === undefined);

    if (emptyCells.length === 0) {
      throw new Error('Board is already filled! There are no empty cells!');
    }

    return emptyCells.reduce((min, cell) => {
      return cell.getCandidates().size < min.getCandidates().size ? cell : min;
    });
  }

  public getCellFromCoordinates(coordinates: Coordinates) {
    if (coordinates.row >= this.unitSize) {
      throw new Error(
        'Requested row is out of bounds for this board! ' +
          `Requested row: ${coordinates.row}, Max row: ${this.unitSize}`
      );
    } else if (coordinates.col >= this.unitSize) {
      throw new Error(
        'Requested column is out of bounds for this board! ' +
          `Requested column: ${coordinates.col}, Max column: ${this.unitSize}`
      );
    }

    // Non-null assertion is safe, we would have already thrown
    return this.cells[coordinates.row]![coordinates.col]!;
  }

  public getUnitsFromCoordinates(coordinates: Coordinates) {
    return this.units.filter((unit) =>
      unit.cellCoords.some(
        (c) => c.row === coordinates.row && c.col === coordinates.col
      )
    );
  }

  public getPeersFromCoordinates(targetCoordinates: Coordinates) {
    const unitsForCell = this.getUnitsFromCoordinates(targetCoordinates);
    const coordinates = unitsForCell.map((unit) => unit.cellCoords).flat();
    const otherCoordinates = coordinates.filter((other) => {
      return (
        other.row !== targetCoordinates.row ||
        other.col !== targetCoordinates.col
      );
    });
    return otherCoordinates.map((coordinates) =>
      this.getCellFromCoordinates(coordinates)
    );
  }

  public removeCandidatesFromPeers(coordinates: Coordinates, value: number) {
    this.getPeersFromCoordinates(coordinates).forEach((cell) =>
      cell.removeCandidate(value)
    );
  }

  public initializeCandidates() {
    this.cells.flat().forEach((cell) => {
      const value = cell.getValue();
      if (value !== undefined) {
        this.removeCandidatesFromPeers(cell.coordinates, value);
      }
    });
  }

  public getCandidateCount() {
    return this.cells
      .flat()
      .reduce(
        (count, cell) =>
          count +
          (cell.getValue() === undefined ? cell.getCandidates().size : 0),
        0
      );
  }

  public jsonPrint() {
    const result: number[][] = [];
    this.cells.forEach((row) => {
      const values = row.map((cell) => cell.getValue() ?? 0);
      result.push(values);
    });
    return JSON.stringify(result);
  }

  public prettyPrint() {
    const rows: string[] = [];
    for (let row = 0; row < this.unitSize; row++) {
      let result = '';
      for (let col = 0; col < this.unitSize; col++) {
        result += this.cells[row]![col]!.print();
        const plusOne = col + 1;
        const notAtEnd = plusOne !== this.unitSize;
        if (notAtEnd) {
          result += ' ';
        }
        if (plusOne % this.boxEdgeSize === 0 && notAtEnd) {
          result += '| ';
        }
      }
      rows.push(result);

      const plusOne = row + 1;
      const notAtEnd = plusOne !== this.unitSize;

      if (plusOne % this.boxEdgeSize === 0 && notAtEnd) {
        rows.push(this.getHorizontalDivider());
      }
    }
    // Extra newline for more square boxes on 100x100 and above
    const separator = this.unitSize.toString().length > 2 ? '\n\n' : '\n';
    return rows.join(separator);
  }

  private getHorizontalDivider() {
    const cellWidth = this.unitSize.toString().length;
    let divider = '';
    for (let i = 0; i < this.boxEdgeSize; i++) {
      const plusOne = i + 1;
      const isMiddle = i !== 0 && plusOne !== this.boxEdgeSize;
      // Add one box worth of divisions
      divider += '-'.repeat(this.boxEdgeSize * (cellWidth + 1));
      // Add one extra to account for the extra space between the vertical divider and the first element
      if (isMiddle) {
        divider += '-';
      }
      if (i + 1 !== this.boxEdgeSize) {
        divider += '+';
      }
    }
    return divider;
  }
}

File: kotlin/src/main/kotlin/primitives/Board.kt
package dev.patientallison.sudoku.primitives

import dev.patientallison.sudoku.HouseType
import dev.patientallison.sudoku.data.Coordinates
import dev.patientallison.sudoku.exceptions.BoardFilledException
import dev.patientallison.sudoku.exceptions.IllegalBoardException
import dev.patientallison.sudoku.isPositiveSquare
import kotlin.math.sqrt

class Board(
    val cells: List<List<Cell>>,
) {
    val boxEdgeSize: Int
    val houseSize: Int
    val totalBoardSize: Int
    val houses: List<House>

    init {
        require(isPositiveSquare(cells.size)) {
            "Height is not a square! Given height: ${cells.size}"
        }

        val width = cells[0].size
        require(isPositiveSquare(width)) { "Width is not a square! Given width: $width" }
        require(cells.size == width) {
            "Width and height are not the same! Board is not a square! Height: ${cells.size} Width: $width"
        }
        require(cells.all { it.size == width }) {
            "All rows must be the same width! Expected: $width"
        }

        boxEdgeSize = sqrt(cells.size.toDouble()).toInt()
        houseSize = cells.size
        totalBoardSize = houseSize * width
        houses = buildHouses()
    }

    private fun buildHouses(): List<House> {
        val allCoords =
            cells.flatMapIndexed { row, rowCells ->
                rowCells.indices.map { col -> Coordinates(row, col) }
            }

        val rowHouses = allCoords.groupBy { it.row }.values.map { House(HouseType.ROW, it) }
        val colHouses = allCoords.groupBy { it.col }.values.map { House(HouseType.COLUMN, it) }
        val boxHouses =
            allCoords.groupBy { (it.row / boxEdgeSize) * boxEdgeSize + (it.col / boxEdgeSize) }
                .values.map { House(HouseType.BOX, it) }

        return listOf(rowHouses, colHouses, boxHouses).flatten()
    }

    /**
     * Clones the board for use in backtracking
     * @returns A cloned board
     */
    fun clone(): Board {
        val clonedCells = cells.map { row -> row.map { it.clone() } }
        return Board(clonedCells)
    }

    /**
     * Returns cells belonging to a House
     * @returns cells belonging to a House
     */
    fun getCellsForHouse(house: House): List<Cell> {
        return house.cellCoordinates.map { cells[it.row][it.col] }
    }

    /**'
     * Validates that the current board state is valid
     * @throws IllegalBoardException A list of board violations
     */
    fun validate() {
        val violations = mutableMapOf<House, List<Cell>>()

        houses.forEach { house ->
            val cells = getCellsForHouse(house)
            val duplicates =
                cells
                    .filter { it.getValue() != null }
                    .groupBy { it.getValue()!! }
                    .filter { it.value.size > 1 }

            duplicates.values.forEach {
                if (it.size > 1) violations[house] = it
            }
        }

        if (violations.isNotEmpty()) {
            val stringifiedViolations =
                violations.entries.map { entry ->
                    val stringifiedCells = entry.value.joinToString(", ") { it.toString() }
                    "{ Violated House: ${entry.key}, Violated Cells: $stringifiedCells"
                }
            throw IllegalBoardException("Board is invalid! Violations: [ ${stringifiedViolations.joinToString(", ")} ]")
        }
    }

    /**
     * Determines if every cell in the board has a value. Does not check validity
     * @returns if the board is already filled
     */
    fun isFilled() = cells.flatten().all { it.getValue() != null }

    /**
     * Determines if the board is already solved
     * @returns If the board is solved or not
     */
    fun isSolved(): Boolean {
        try {
            validate()
        } catch (_: IllegalBoardException) {
            return false
        }

        return isFilled()
    }

    /**
     * Get the empty cell with the fewest candidates (the best cell for use in backtracking)
     * @returns the empty cell with the fewest candidates
     * @throws BoardFilledException if the board is already filled and there are no empty cells
     */
    fun selectEmptyCell(): Cell {
        val emptyCells = cells.flatten().filter { it.getValue() == null }

        if (emptyCells.isEmpty()) {
            throw BoardFilledException("Board is already filled! There are no empty cells!")
        }

        return emptyCells.minBy { it.getCandidates().size }
    }

    /**
     * Get a cell from Coordinates
     * @return Cell for the requested Coordinates
     * @throws IllegalArgumentException if Coordinates are out of bounds for the board
     */
    fun getCellFromCoordinates(coordinates: Coordinates): Cell {
        require(coordinates.row < houseSize) {
            "Requested row is out of bounds for this board! Requested row: ${coordinates.row}, Max row: $houseSize"
        }

        require(coordinates.col < houseSize) {
            "Requested column is out of bounds for this board! Requested column: ${coordinates.col}, Max column: $houseSize"
        }

        return cells[coordinates.row][coordinates.col]
    }

    fun getHousesFromCoordinates(coordinates: Coordinates): List<House> {
        return houses.filter { house ->
            house.cellCoordinates.any { it.row == coordinates.row && it.col == coordinates.col }
        }
    }

    fun getPeersFromCoordinates(coordinates: Coordinates): List<Cell> {
        val housesForCell = getHousesFromCoordinates(coordinates)
        val houseCoordinates = housesForCell.flatMap { it.cellCoordinates }
        val otherCoordinates = houseCoordinates.filterNot { it == coordinates }.distinct()
        return otherCoordinates.map { getCellFromCoordinates(it) }
    }

    fun removeCandidatesFromPeers(
        coordinates: Coordinates,
        value: Int,
    ) {
        getPeersFromCoordinates(coordinates).forEach { it.removeCandidate(value) }
    }

    fun initializeCandidates() {
        this.cells.flatten().forEach {
            val value = it.getValue()
            if (value != null) removeCandidatesFromPeers(it.coordinates, value)
        }
    }

    fun getCandidateCount() =
        cells.flatten().sumOf {
            if (it.getValue() == null) it.getCandidates().size else 0
        }

    fun prettyPrint(): String {
        val rows = mutableListOf<String>()
        (0 until houseSize).forEach { row ->
            val cellsList = mutableListOf<String>()
            (0 until houseSize).forEach { col ->
                val plusOne = col + 1
                val notAtEnd = plusOne != houseSize
                var cellString = cells[row][col].print()
                if (notAtEnd) cellString += " "
                if (plusOne % boxEdgeSize == 0 && notAtEnd) cellString += "| "
                cellsList.add(cellString)
            }
            rows.add(cellsList.joinToString(""))
            val plusOne = row + 1
            val notAtEnd = plusOne != houseSize
            if (plusOne % boxEdgeSize == 0 && notAtEnd) {
                rows.add(getHorizontalDivider())
            }
        }

        val separator = if (houseSize.toString().length > 2) "\n\n" else "\n"
        return rows.joinToString(separator)
    }

    private fun getHorizontalDivider(): String {
        val cellWidth = houseSize.toString().length
        val divider = mutableListOf<String>()
        (0 until boxEdgeSize).forEach {
            val plusOne = it + 1
            val isMiddle = it != 0 && plusOne != boxEdgeSize
            // Add one box worth of divisions
            var boxDivider = "-".repeat(boxEdgeSize * (cellWidth + 1))
            // Add one extra to account for the extra space between the vertical divider and the first element
            if (isMiddle) boxDivider += "-"
            if (plusOne != boxEdgeSize) boxDivider += "+"
            divider.add(boxDivider)
        }
        return divider.joinToString("")
    }
}
